### PR TITLE
[SPARK-43199][SQL] Make InlineCTE idempotent

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR fixes `InlineCTE`'s idempotence. E.g. the following query:
```
WITH
  x(r) AS (SELECT random()),
  y(r) AS (SELECT * FROM x),
  z(r) AS (SELECT * FROM x)
SELECT * FROM z
```
currently breaks it because we take into account the reference to `x` from `y` when deciding about not inlining `x` in the first round:
```
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.InlineCTE ===
 WithCTE                                                        WithCTE
 :- CTERelationDef 0, false                                     :- CTERelationDef 0, false
 :  +- Project [rand()#218 AS r#219]                            :  +- Project [rand()#218 AS r#219]
 :     +- Project [random(2957388522017368375) AS rand()#218]   :     +- Project [random(2957388522017368375) AS rand()#218]
 :        +- OneRowRelation                                     :        +- OneRowRelation
!:- CTERelationDef 1, false                                     +- Project [r#222]
!:  +- Project [r#219 AS r#221]                                    +- Project [r#220 AS r#222]
!:     +- Project [r#219]                                             +- Project [r#220]
!:        +- CTERelationRef 0, true, [r#219]                             +- CTERelationRef 0, true, [r#220]
!:- CTERelationDef 2, false                                     
!:  +- Project [r#220 AS r#222]                                 
!:     +- Project [r#220]                                       
!:        +- CTERelationRef 0, true, [r#220]                    
!+- Project [r#222]                                             
!   +- CTERelationRef 2, true, [r#222]    
```
But in the next round we inline `x` because `y` was removed due to lack of references:
```
Once strategy's idempotence is broken for batch Inline CTE
!WithCTE                                                        Project [r#222]
!:- CTERelationDef 0, false                                     +- Project [r#220 AS r#222]
!:  +- Project [rand()#218 AS r#219]                               +- Project [r#220]
!:     +- Project [random(2957388522017368375) AS rand()#218]         +- Project [r#225 AS r#220]
!:        +- OneRowRelation                                              +- Project [rand()#218 AS r#225]
!+- Project [r#222]                                                         +- Project [random(2957388522017368375) AS rand()#218]
!   +- Project [r#220 AS r#222]                                                +- OneRowRelation
!      +- Project [r#220]                                       
!         +- CTERelationRef 0, true, [r#220]    
```

### Why are the changes needed?
We use `InlineCTE` as an idempotent rule in the `Optimizer`, `CheckAnalysis` and `ProgressReporter`.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @cloud-fan, @maryannxue

**Comment by peter-toth** [comment]
> @peter-toth can you briefly explain the idea of fixing it?

I've updated the PR recently, but the main change is that the CTE accumulator map argument of `buildCTEMap()` changed from 
`mutable.HashMap.empty[Long, (CTERelationDef, Int)]`
to 
`mutable.SortedMap.empty[Long, (CTERelationDef, Int, mutable.Map[Long, Int])]`.
and the new `mutable.Map[Long, Int]` part tracks where the references are pointing to from a CTE. (The old `Int` part tracks the "count of incoming references".)

Once we have this extended outer we can correct the "count of incoming references" in `cleanCTEMap()`. We just need to iterate the keys in reverse order (that's why the outer map is now a `SortedMap`) and if we encounter a CTE whose "count of incoming references" is 0 then we decrease the referenced CTE's "count of incoming references".

To build the new inner map `buildCTEMap` has a new `outerRefMap` optional argument.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala [hunk 1]
```diff
@@ -42,8 +42,9 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     if (!plan.isInstanceOf[Subquery] && plan.containsPattern(CTE)) {
-      val cteMap = mutable.HashMap.empty[Long, (CTERelationDef, Int)]
+      val cteMap = mutable.SortedMap.empty[Long, (CTERelationDef, Int, mutable.Map[Long, Int])]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:45]: this map is very complicated now, let's add some comments to explain what key/value means, including the inner map.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala [hunk 2]
```diff
@@ -70,48 +71,69 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
...
       case _ =>
-    }
-
-    if (plan.containsPattern(CTE)) {
-      plan.children.foreach { child =>
-        buildCTEMap(child, cteMap)
-      }
+        if (plan.containsPattern(CTE)) {
+          plan.children.foreach { child =>
+            buildCTEMap(child, cteMap, outerCTEId)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:97]: is this duplicated? If plan is `WithCTE`, we should have already invoked `buildCTEMap` for CTE relations in https://github.com/apache/spark/pull/40856/files#diff-1c15413e5d63f78fff1db3dec9df4a671e78b76d086104d81f4a967eb2800805R82

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:97]: nvm, this is the `case _` branch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala [hunk 3]
```diff
@@ -70,48 +71,69 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
...
           }
         }
+    }
+  }
+
+  private def cleanCTEMap(
+      cteRefMap: mutable.SortedMap[Long, (CTERelationDef, Int, mutable.Map[Long, Int])]
+    ) = {
+    cteRefMap.keys.toSeq.reverse.foreach { currentCTEId =>
+      val (_, currentRefCount, refMap) = cteRefMap(currentCTEId)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:116]: so the idea is, if a CTE relation A is referenced by another CTE relation B, and relation B has no references, we should update the relation A reference count to exclude the references from relation B? Can we add comments to write it down?

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala]: Yes. Added scaladoc in https://github.com/apache/spark/pull/40856/commits/ecbc4b879a8f1c397d200958b1b7ade0f6768ca5, let me know if it needs more detailed comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala [hunk 4]
```diff
@@ -68,50 +69,91 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
...
 
+  /**
+   * Accumulates all the CTEs from a plan into a special map.
+   *
+   * @param plan The plan to collect the CTEs from
+   * @param cteMap A mutable map that accumulates the CTEs and their reference information by CTE
+   *               ids. The value of the map is tuple whose elements are:
+   *               - The CTE definition
+   *               - The number of incoming references to the CTE. This includes references from
+   *                 outer CTEs and regular places.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:80]:
```suggestion
   *                 inner CTEs and regular places.
```

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala (hunk 4 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:80]: I actually wanted to write `other CTEs` and not inner/outer. E.g. in
```
WITH (
  cte1 AS (SELECT 1),
  cte2 AS (SELECT * FROM cte1)
)
SELECT * FROM cte1 JOIN cte2 ON ...
```
the reference count of cte1 is 2. 1 is from an "other" CTE (but not "inner"/"outer"), 1 is from a "regular place".

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala]: fixed in https://github.com/apache/spark/pull/40856/commits/806c2de2b7db9b960c591a849dc43be7f65b468b

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala]: @cloud-fan, please let me know if anything else is needed here.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by peter-toth** [comment]
Thanks @cloud-fan!

**Pull request closed without merging**